vochain/ist/ist.go

```diff
@@ -0,0 +1,303 @@
+// Package ist provides the Internal State Transition (IST) mechanism for Vochain.
+// The IST is responsible for modifying the Vochain's state based on actions scheduled
+// to be executed at specific heights. This ensures accurate, tamper-proof results
+// through consensus among nodes.
+//
+// The Internal State Transition Controller (ISTC) is the main component of this package.
+// It schedules and executes IST actions at specified heights. Actions include computing
+// results, committing results, and ending processes.
+//
+// The package contains the following types:
+//
+// - Controller: the IST controller.
+// - Actions: a model to store the list of IST actions for a specific height.
+// - Action: a model to store the IST actions.
+//
+// IST actions are encoded and decoded using the Gob encoding package.
+// The encoding and decoding functions are provided by the Actions type.
+//
+// The Controller struct provides methods for scheduling, rescheduling, and committing
+// IST actions. When committing an action, it performs the necessary operations depending
+// on the type of action, such as computing results, committing results, or ending processes.
+// It also provides methods to store and retrieve data from the no-state, which is a separate
+// key-value store that holds non-consensus-critical data.
+//
+// Constants used in the package are:
+//
+// - BlocksToWaitForResultsFactor: a factor used to compute the height at which the results will be committed to the state.
+// - ExtraWaitSecondsForResults: the number of extra seconds to wait for results.
+//
+// Technical decisions taken in the code include:
+//
+// - Implementing a separate no-state key-value store for non-consensus-critical data storage.
+// - Computing results in a separate goroutine to avoid blocking state transitions.
+// - Including a mechanism for rescheduling actions in case of missing keys or other issues.
+package ist
+
+import (
+	"bytes"
+	"encoding/gob"
+	"encoding/hex"
+	"errors"
+	"fmt"
+
+	"go.vocdoni.io/dvote/db"
+	"go.vocdoni.io/dvote/log"
+	"go.vocdoni.io/dvote/vochain/results"
+	"go.vocdoni.io/dvote/vochain/state"
+	"go.vocdoni.io/proto/build/go/models"
+)
+
+const (
+	dbPrefix = "ist/"
+)
+
+// Controller is the internal state transition controller.
+type Controller struct {
+	st              *state.State
+	isSynchronizing uint32
+}
+
+// NewISTC creates a new ISTC.
+// The ISTC is the controller of the internal state transitions.
+// It schedule actions to be executed at a specific height that will modify the state.
+func NewISTC(s *state.State) *Controller {
+	return &Controller{
+		st: s,
+	}
+}
+
+const (
+	// ActionComputeResults computes and schedules the commit of results, for the given election and height.
+	ActionComputeResults = iota
+	// ActionCommitResults schedules the commit of the results to the state.
+	ActionCommitResults
+	// ActionEndProcess sets a process as ended. It schedules ActionComputeResults.
+	ActionEndProcess
```

Comment:
make these typed constants, to ensure that you don't mix them with other integer values